### PR TITLE
Issue #5556: CUDA-device determinism path coverage (cheap-lane worker)

### LINKED ISSUE #5556: bug: make Torch 2.13.0 safe on Python 3.12 CI and smoke paths
## Goal / Problem

**Observed behavior**

The Torch 2.13.0 dependency graph fails the repository's required CI/smoke contract on the
Dependabot branches for PRs #5553 and #5551. After the required branch refreshes:

- PR #5553 head `6255754f48dad30ce5ab00ee5f71150eeb16a525` has failed required
  `fast-feedback` and `smoke-artifacts`/`xdist-scratch-isolation` jobs.
- PR #5551 head `0b9f57975b3681739ebdbe6cc9b20d792e3ebe84` has a failed required
  `smoke-artifacts` job.
- The completed predecessor run for PR #5553, run
  [29310500779](https://github.com/ll7/robot_sf_ll7/actions/runs/29310500779), records
  `Fatal Python error: Segmentation fault` in the Python 3.12 test workers while importing
  `torch._dynamo`/`torch.utils._triton` through `torch.use_deterministic_algorithms` in
  `tests/conftest.py`; the example smoke jobs also exit with code `-11`.

This is a runtime compatibility regression, not an advisory or prose-only review concern.

**Expected behavior**

Torch 2.13.0 and its CUDA extra must install and pass the configured CPU/Python 3.12,
Python 3.13, GPU-extra, unit-test, xdist, artifact-smoke, and example-smoke paths.

## Archetype Metadata

```yaml
archetype: implementation
evidence_tier: blocked
linked_policy:
  - docs/code_review.md
```

## Reproduction

Use either affected PR head and the repository's canonical dependency/test paths:

```bash
uv sync --locked --all-extras
uv run pytest -n auto
```

CI evidence:

- [PR #5553 current CI](https://github.com/ll7/robot_sf_ll7/actions/runs/29311676700)
- [PR #5551 current CI](https://github.com/ll7/robot_sf_ll7/actions/runs/29311742809)
- [completed Torch 2.13 failure log](https://github.com/ll7/robot_sf_ll7/actions/runs/29310500779)

## Scope

- Affected module(s): Torch import/determinism compatibility in `tests/conftest.py` and
  `robot_sf/common/seed.py`, plus any minimal CI/environment contract needed to prove it.
- Regression: dependency resolution to Torch 2.13.0 in PRs #5553 and #5551.
- In scope: identify the crash mechanism, implement the smallest safe compatibility fix or
  dependency constraint, and add a regression check covering the failing path.
- Out of scope: unrelated planner behavior, broad test-suite refactors, or a separate Torch
  feature upgrade.

## Added Value Estimation

- User impact: high; affected installs cannot complete required tests or smoke examples.
- Benchmark impact: high; benchmark/training environments using the updated Torch graph are not
  currently trustworthy.
- Maintenance impact: medium; the fix should make the dependency contract explicit and testable.
- Why now: this gate cannot merge either Torch 2.13 dependency PR while required checks fail.

## Effort Estimation

- Rough estimate (hours): 4
- Best estimate (hours): 8
- Unknowns: whether the segfault is fixed by a narrow determinism-import guard or requires a
  Torch version exclusion/upgrade.

## Complexity Estimation

- Implementation complexity: medium
- Dependencies: Torch 2.13.0, CUDA 13 dependency graph, Python 3.12 CI, xdist, and smoke examples.
- Open questions: exact failing native boundary and whether CPU-only and CUDA-extra paths share it.

## Risk Assessment

- Functional risk: high until the crash is removed; a workaround must not silently disable the
  repository's determinism contract.
- Compatibility risk: high across Python 3.12/3.13 and CPU/GPU environments.
- Rollout risk: medium; dependency changes can invalidate benchmark comparability or local caches.
- Mitigation: fail closed on the affected CI paths, reproduce on both extras, preserve an explicit
  determinism assertion, and require the full configured gate before promoting the dependency PRs.

## Affected Files

- `tests/conftest.py` - current CI stack enters Torch through the autouse determinism fixture.
- `robot_sf/common/seed.py` - runtime seed/determinism path uses the same Torch API.
- `pyproject.toml` / `uv.lock` - dependency constraints and resolved Torch/CUDA graph.

## Definition of Done

- [ ] Root cause of the Torch 2.13.0 crash is identified with a minimal reproduction.
- [ ] A minimal compatibility fix or safe dependency constraint is implemented.
- [ ] A regression test covers the failing determinism/import path.
- [ ] Python 3.12 CPU and Python 3.13 paths pass.
- [ ] The CUDA extra path passes on its supported runner/host.
- [ ] `fast-feedback`, `smoke-artifacts`, `xdist-scratch-isolation`, and `examples-smoke` pass
      on the affected dependency PR head(s).

## Success Metrics

- No segmentation fault or exit code `-11` in the affected CI jobs.
- Full PR checks complete successfully for both the training and CUDA-extra dependency paths.
- Determinism behavior remains explicit and covered rather than disabled silently.

## Validation / Testing

- [ ] Run the focused Python 3.12 reproduction.
- [ ] Run the focused Torch determinism/seed tests.
- [ ] Run the repository PR readiness gate on the fix.
- [ ] Re-run the affected PR CI and confirm all required checks are green.

## Estimate Discussion

- Why these values were proposed: the failure is reproducible in hosted CI but the native crash
  boundary still needs isolation.
- Uncertainty / confidence: 75% confidence that the two PRs share the same Torch 2.13 compatibility
  root cause; current-job logs are still completing.
- What evidence would move the estimate: a Python 3.12 minimal reproducer or a clean full CI rerun.

## Project Metadata

- Priority: P1
- Effort (h): 4-8
- Reviewed: 2026-07-14 gate review

### PR DESCRIPTION
**Pull request by ll7**
## What / Why

Issue #5556 was closed after supported-host CUDA validation completed. This PR adds a small post-closure hardening slice: two in-tree tests that keep the already-merged Torch 2.13 determinism fix executable on future CUDA-capable test hosts.

This is regression-test coverage, not new evidence required to close #5556 and not a replacement for its recorded supported-host validation.

## Changes

- `test_torch_213_determinism_guard_runs_on_cuda_device` runs `set_global_seed` on CUDA and verifies that deterministic algorithms and the cuDNN deterministic flag are enabled without entering the former crashing public wrapper path.
- `test_torch_213_cuda_seed_is_reproducible_across_calls` verifies that two CUDA tensors generated after identical seeds are byte-equal.
- Both tests skip explicitly when CUDA is unavailable, so CPU-only runners retain their existing contract.

## Validation / Proof

Gate-verified on 2026-07-16:

- Python 3.13 worktree environment
- Torch `2.13.0+cu130`
- CUDA runtime 13.0 available
- NVIDIA GeForce RTX 3070
- `scripts/dev/run_focused_tests.sh tests/test_seed_utils.py -q` — **14 passed**, including both CUDA tests
- `uv run ruff check tests/test_seed_utils.py` — passed
- `uv run ruff format --check tests/test_seed_utils.py` — passed

Claim boundary: this proves the in-tree CUDA regression tests execute on the gate host. It does not create benchmark evidence or reopen #5556's completed supported-host validation.

## Successor statement

This slice does not duplicate PRs #5577, #5652, #5685, or #5767, which added the runtime helpers, dependency contract, and CPU/mocked regression coverage. It preserves a direct real-device regression check after #5556's closure.

## Exact sentence

This PR was produced by the OpenGateway/auto-smart-routing cheap implementation lane; the review gate hardens and merges.

Refs #5556

### CODE REVIEW

**Comment by ll7** [comment]
Gate findings for `feacd6adf0277ec59464d18a7f9358e721138a88` (phase 1 review of the actual diff):

1. [MEDIUM, open Gemini thread](https://github.com/ll7/robot_sf_ll7/pull/5861#discussion_r3596080765): `test_torch_213_cuda_seed_is_reproducible_across_calls` dereferences the dynamically imported module without the explicit non-`None` guard already used by the sibling CUDA test. Verdict: **fixed in phase 2 required** with the defensive assertion before CUDA calls.
2. [Body/provenance contract] the live body says this PR supplies #5556's remaining CUDA DoD and that hosted-GPU `examples-smoke` remains outstanding, but #5556 was subsequently closed by the maintainer after supported-host CUDA validation. The body also attributes execution to an RTX 3080 without a durable artifact link. Verdict: **fixed in phase 2 required** by reframing this as post-closure in-tree regression hardening, removing the obsolete blocker language, and recording only gate-verified hardware/version/test evidence.
3. [Freshness checklist] current `origin/main` is `172a2c0438803129ed2eb621a767f75481eeb8f5`, but the reviewed head does not contain it. Verdict: **fixed in phase 2 required** by updating the branch and obtaining green CI on the new exact head.

Open-review-thread reconciliation: 1 open thread, enumerated as finding 1 above. CodeRabbit was rate-limited and supplied no review findings; this gate is the review of record. No other diff findings. Intended-design classification: intentionally narrowed post-closure GPU regression-test hardening, blocked only on the guard, truthful live-body provenance, and current-main proof.

gate-verdict: changes-requested @ feacd6adf0277ec59464d18a7f9358e721138a88

**Comment by ll7** [comment]
Final gate findings for `9abe63f8f9185939d74ce12703d062b28c65e37b`:

1. **Gemini MEDIUM thread — missing non-`None` Torch guard before CUDA calls:** fixed with an explicit assertion. Thread resolved.
2. **Body/provenance drift — #5556 had already closed and the body described stale remaining work/unsupported hardware:** fixed. The current body truthfully scopes this as post-closure regression hardening, records the gate host (Torch `2.13.0+cu130`, CUDA 13.0, RTX 3070), and uses `Refs #5556`.
3. **Stale-main admission:** fixed. Current main `a08484b54d5a48dd4da0e4e991309b29078b4221` is an ancestor of this exact head.

0 open review threads (1 total review-bot thread, resolved).

Focused validation passed **14 tests**, including both real-device CUDA tests; changed-file Ruff/format checks pass. Exact-head CI is **24/24 green**. The earlier full local readiness core lane passed and stopped only on main-side TODO-docstring ratchet debt outside this one-file branch; #5858 tracks the stale `tests/conftest.py` baseline and #5867 tracks `tests/validation/test_pr_contract_check.py`. This diff adds no placeholder docstring. Exclusions are clear and the exact-SHA batch-gate admission is allowed.

gate-verdict: accepted @ 9abe63f8f9185939d74ce12703d062b28c65e37b
